pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -354,6 +362,71 @@ impl WalreceiverState {
         });
     }
 
+    /// Drops the current connection (if any) and updates retry timeout for the next
+    /// connection attempt to the same safekeeper.
+    async fn drop_old_connection(&mut self, needs_shutdown: bool) {
+        let wal_connection = match self.wal_connection.take() {
+            Some(wal_connection) => wal_connection,
+            None => return,
+        };
+
+        if needs_shutdown {
```

Comment:
NIT: this looks a bit dangerous, since unattended walreceiver connection is not what we want generally.
`drop_old_connection(false)` is only called on the task end and boils down to `TaskHandle`:
```rust
async fn shutdown(self) {
    self.cancellation.send(()).ok();
    if let Err(e) = self.handle.await {
        error!("Task failed to shut down: {e}")
    }
}
```

to me, this code is safer to run twice than never.
I wonder, should we put it the code above into `Drop` impl.

Reply:
Fixed, looks safe to call twice, but I wonder what will happen if `self.handle.await` is called twice?

Reply:
just tried,
`thread 'main' panicked at 'JoinHandle polled after completion', /home/ars/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.12.0/src/runtime/task/core.rs:208:22
`

Reply:
It should return the same result as it returned on the first invocation, at least that's what I've noted in my experiments and there could be some exceptional situations.

Reply:
Looks like I was wrong, we can try https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.is_finished then.

Reply:
`is_finished` returns true if the task was finished, not if the result was consumed. I see no better way than reverting to `needs_shutdown`.

Reply:
Well, I'd just maintain 'if we awaited, then `self.wal_connection` is None'; so we always await if there is a handle, even if we know through `TaskEvent::End` that task is going to finish.